patches/tModLoader/Terraria/ModLoader/Core/ModCompile.cs

```diff
@@ -56,18 +56,38 @@ public BuildingMod(TmodFile modFile, BuildProperties properties, string path) :
 
 	internal static string[] FindModSources()
 	{
-		Directory.CreateDirectory(ModSourcePath);
-		return Directory.GetDirectories(ModSourcePath, "*", SearchOption.TopDirectoryOnly).Where(dir => {
-			var directory = new DirectoryInfo(dir);
-			return directory.Name[0] != '.' && directory.Name != "ModAssemblies" && directory.Name != "Mod Libraries";
-		}).ToArray();
+		var modSources = new List<string>();
+
+		// Find any mod sources defined in the ModSources directory.
+		if (Directory.Exists(ModSourcePath)) {
+			modSources.AddRange(Directory.GetDirectories(ModSourcePath, "*", SearchOption.TopDirectoryOnly).Where(dir => {
+				var directory = new DirectoryInfo(dir);
+				return directory.Name[0] != '.' && directory.Name != "ModAssemblies" && directory.Name != "Mod Libraries";
+			}));
+		}
+
+		// Find mod sources defined by built .tmod files.
+		// It's possible for AllFoundMods to not be populated in low-init scenarios
+		// such as mod building, so we can populate it ourselves.
+		var foundMods = ModOrganizer.AllFoundMods ?? ModOrganizer.FindAllMods();
+		modSources.AddRange(ModOrganizer.AllFoundMods.Where(m => m.location == ModLocation.Local).Select(m => m.properties.modSource).Where(s => !string.IsNullOrEmpty(s)));
+
+		return modSources.Distinct().Where(Directory.Exists).ToArray();
+	}
+
+	internal static string GetModSource(TmodFile tmod)
+	{
+		var buildInfo = BuildProperties.ReadModFile(tmod);
+		return string.IsNullOrEmpty(buildInfo.modSource) ? Path.Combine(ModSourcePath, tmod.Name) : buildInfo.modSource;
 	}
 
+	internal static string GetModSource(LocalMod mod) => string.IsNullOrEmpty(mod.properties.modSource) ? Path.Combine(ModSourcePath, mod.Name) : mod.properties.modSource;
+
```

Comment:
Do we need this at all? It would only affect modders trying to publish a mod they had built on a prior tML update. A rebuild of the mod would ensure it has source path metadata